src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs

```diff
@@ -0,0 +1,302 @@
+        /// Gets an asynchronous package core reader.
+        /// </summary>
+        /// <exception cref="ObjectDisposedException">Thrown if this object is disposed.</exception>
+        public IAsyncPackageCoreReader CoreReader
+        {
+            get
+            {
+                ThrowIfDisposed();
+
+                return _packageReader.Value;
+            }
+        }
+
+        public ISignedPackageReader SignedPackageReader
+        {
+            get
+            {
+                ThrowIfDisposed();
+
+                return _packageReader.Value;
+            }
+        }
+
+        public string Source { get; }
+
+        /// <summary>
+        /// Initializes a new <see cref="LocalPackageArchiveDownloader" /> class.
+        /// </summary>
+        /// <param name="source">A package source.</param>
+        /// <param name="packageFilePath">A source package archive file path.</param>
+        /// <param name="packageIdentity">A package identity.</param>
+        /// <param name="logger">A logger.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="packageFilePath" />
+        /// is either <c>null</c> or an empty string.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="packageIdentity" />
+        /// is either <c>null</c> or an empty string.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="logger" />
+        /// is either <c>null</c> or an empty string.</exception>
+        public LocalPackageArchiveDownloader(
+            string source,
+            string packageFilePath,
+            PackageIdentity packageIdentity,
+            ILogger logger)
+        {
+            if (string.IsNullOrEmpty(packageFilePath))
+            {
+                throw new ArgumentException(
+                    string.Format(CultureInfo.CurrentCulture, Strings.StringCannotBeNullOrEmpty, nameof(packageFilePath)),
+                    nameof(packageFilePath));
+            }
+
+            if (packageIdentity == null)
+            {
+                throw new ArgumentNullException(nameof(packageIdentity));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            _packageFilePath = packageFilePath;
+            _packageIdentity = packageIdentity;
+            _logger = logger;
+            _packageReader = new Lazy<PackageArchiveReader>(GetPackageReader);
+            _sourceStream = new Lazy<FileStream>(GetSourceStream);
+            _handleExceptionAsync = exception => Task.FromResult(false);
+            Source = source;
+        }
+
+        /// <summary>
+        /// Disposes of this instance.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!_isDisposed)
+            {
+                if (_packageReader.IsValueCreated)
+                {
+                    _packageReader.Value.Dispose();
```

Comment:
When will this get disposed if IsValueCreated == false ?

Reply:
If it wasn't created, there's nothing to dispose.
This check is good because otherwise you'd initialize and dispose of the object.